### ISSUE TITLE
make validating `Recursor` cache intermediate DNSSEC validations

### ISSUE THREAD

**Issue by japaric** [hickory-dns/hickory-dns#2294]
as it was noted in the description of PR #2289 , the `Recursor` does *not* cache the results of each step of the chain of trust validation process. if it did, the `Recursor` could avoid performing signature validations (crypto operations) on subsequent queries about sibling and parent domains.

This behavior can be observed in the `explore` example

First query: `SOA nameservers.com.`

each line of the form "validated ({name}, {record_type}) with " in the logs indicates a RRSIG validation (crypto) operation

``` console
validated (com., DS) with (..)
validated (nameservers.com., DS) with (..)
validated (nameservers.com., SOA) with (..)
```

Second query: `DS nameservers.com.`

``` console
validated (com., DS) with (..)
validated (nameservers.com., DS) with (..)
```

There it can be seen that two RRSIG validation operations, that colud have been avoided, where repeated in the second query

**Comment by japaric** [comment]
actually, it seems that the whole chain of trust validation happens (+) even happens even when the exact same query is repeated  :thinking: 

(+) the network operations are not repeated; we have tests for that

**Comment by bluejekyll** [comment]
Was this fixed in #2297?

**Comment by japaric** [comment]
@bluejekyll no, #2297 only caches the `Proof` outcome on the "top-most" answer that will be returned to the client. the goal of this ticket is to extend caching of `Proof`s to *each* step of the chain of trust validation.

**Comment by cpu** [comment]
I believe this is addressed by https://github.com/hickory-dns/hickory-dns/pull/3351, please reopen with details on what's missing if I'm wrong!
